### PR TITLE
Fix memory leak in inabox-viewport-friendly introduced by inabox-lite.

### PR DESCRIPTION
**Pull request by zombifier**
The leak is caused by InaboxViewportImpl not disconnecting the binding on visibilityChange=hidden, so when the inabox iframe is cleared the reference to it remains in the top window. Readd code for handling the event, with a workaround for Safari not firing it on unload.

### CODE REVIEW

File: src/inabox/inabox-viewport.js [hunk 1]
```diff
@@ -467,6 +467,13 @@ export class ViewportBindingInabox {
     /** @private {?UnlistenDef} */
     this.unobserveFunction_ = null;
 
+    const unloadListener = this.win.addEventListener('unload', () => {
```
Comment (lannka) [inline src/inabox/inabox-viewport.js]: is it guaranteed to fire `visibilitychange` on window `unload`?

if so, we might want to fix the `visibilitychange` regression. and we will need to make sure inaboxViewportBinding has idempotent `connect()` `disconnect()`.

Comment (zombifier) [reply src/inabox/inabox-viewport.js]: It looks like the case: https://w3c.github.io/page-visibility/ (though [Safari might behave differently](https://bugs.webkit.org/show_bug.cgi?id=151234)?

Making `visibilityChange` fire again does seem like the solution. I'll fix it, test it on Safari and get back to you once it's done.

File: src/inabox/inabox-viewport.js [hunk 2]
```diff
@@ -498,13 +520,11 @@ export class ViewportBindingInabox {
...
     // be available immediately).
     // TODO(lannka): Investigate why this is the case.
-    if (this.unobserveFunction_) {
-      return Promise.resolve();
-    }
     return Services.resourcesPromiseForDoc(
       this.win.document.documentElement
     ).then(() => {
-      this.unobserveFunction_ = this.topWindowPositionObserver_.observe(
+      this.unobserveFunction_ = this.unobserveFunction ||
```
Comment (lannka) [inline src/inabox/inabox-viewport.js]: typo, missing `_`

Comment (zombifier) [reply src/inabox/inabox-viewport.js]: Whoops, done!

File: src/inabox/inabox-viewport.js (hunk 1 above)
Comment (zombifier) [reply src/inabox/inabox-viewport.js]: Unfortunately from testing looks like Safari does not fire `visibilityChange` on `unload`, so this fix doesn't work there.

Comment (zombifier) [reply src/inabox/inabox-viewport.js]: The next thing to try is probably the `pagehide` event, which I think is acceptable?

Comment (lannka) [reply src/inabox/inabox-viewport.js]: SGTM.

does that mean the issue existed in Safari before inabox-lite?

File: src/inabox/inabox-viewport.js [hunk 3]
```diff
@@ -133,10 +133,21 @@ class InaboxViewportImpl {
...
     this.binding_.onResize(this.resize_.bind(this));
 
+    /** @private {boolean} */
+    this.visible_ = false;
+    this.ampdoc.onVisibilityChanged(this.updateVisibility_.bind(this));
+    this.updateVisibility_();
+
+    // Workaround for Safari not firing visibilityChange when the page is
+    // unloaded (https://bugs.webkit.org/show_bug.cgi?id=151234).
+    const unloadListener = win.addEventListener('pagehide', () => {
```
Comment (lannka) [inline src/inabox/inabox-viewport.js]: @dvoytenko is `ampdoc` not handling this event?

Comment (lannka) [reply src/inabox/inabox-viewport.js]: https://github.com/ampproject/amphtml/blob/1c5d54b5e3e4a2c108a558fa66b9cc22ca61058e/src/utils/document-visibility.js#L87-L93

Comment (dvoytenko) [reply src/inabox/inabox-viewport.js]: Not to my knowledge at this time. This is a newer event I believe. Not sure how it interops with `visibilityChange` besides unloading nuance. But generally it sounds possible given that the new page visibility events are more clear on the spec side for unload/freeze/terminate lifecycle. I'd be totally ok if we decided to handle it on the `ampdoc` side.

Comment (lannka) [reply src/inabox/inabox-viewport.js]: SG. @zombifier to not block your work here, please add a TODO that this event handling can be moved to ampdoc later.

Comment (zombifier) [reply src/inabox/inabox-viewport.js]: Done.

**Comment by lannka** [comment]
Good to go! Thanks for the fix.

**Comment by zombifier** [comment]
Tested on Safari, it works now. I wanted to add unit tests but we don't have tests set up for InaboxViewportImpl yet, so that can be a TODO. In the meantime are there any other approvals we need?

**Comment by Bnaya** [comment]
Hey I've encountered periodic memory leak in my site, related to the `inabox-viewport-friendly` experiment i believe.
Is there a way i can force this experiment for a page for testing proposes?
(I'm coming from the publisher side, not advertiser)

I will throw in some more info on the leak i've found:
I've debugged minified code.
We are doing ads refresh on spa navigate, using DFP.
When the leak kicks in, i've observed the following: 
`window.ampInaboxPositionObserver` is an instance named `nk`
I've tracked it down to:
Screen Shot 2020-02-06 at 9 49 00

Screen Shot 2020-02-06 at 9 40 54
in `https://cdn.ampproject.org/rtv/012001251659540/amp4ads-v0.js:formatted`
AFSIK When the instance name is nk, it means it was created inside the iframe and not the host frame.
**The leak itself**
Using chrome heap snapshot i've figured,
when `ampInaboxPositionObserver` is `nk`,
`window.ampInaboxPositionObserver.Nb.M` is the array the holds references to functions that holds refs to all the iframes.
After each ads refresh, this array got bigger and bigger, same for the heap size.
When setting this array to `[]` and hitting force GC, the memory consumption dropped. 
IANM, this array holds disposers to intersection observers related to each frame.

Since Yesterday, but before your release, i wasn't able to get to that state again, but i suspect it still effects my users.

Thanks!

**Comment by zombifier** [comment]
Hi @Bnaya, that was indeed the memory leak I found and hopefully fixed in this PR. The new release containing the fix is not released yet, but we expect it to come out soon.

To force the experiment you can open the console in an AMP ad iframe, and run `AMP.toggleExperiment('inabox-viewport-friendly');`. It should apply to all ad slots on your page. Thanks and sorry for the trouble!

**Comment by Bnaya** [comment]
Thanks for your response!
Will `AMP.toggleExperiment('inabox-viewport-friendly');` work for future refreshes of the page?
Because suspect it must run before the ad host

**Comment by Bnaya** [comment]
Also, what's the % of users that expected to get the experiment?

**Comment by Bnaya** [comment]
Another question:  is there a way to ensure out users won't get that experiment?
Given we only control the top frame, and we don't want them to get it even on the first run

**Comment by zombifier** [comment]
> Will AMP.toggleExperiment('inabox-viewport-friendly'); work for future refreshes of the page?
Yes it should.
> Another question: is there a way to ensure out users won't get that experiment?
It has been turned off for now. Once again sorry for the trouble!

**Pull request merged**